src/dispatching/dialogue/mod.rs

```diff
@@ -179,3 +171,130 @@ where
         self.storage.clone().remove_dialogue(self.chat_id).await
     }
 }
+
+/// Perform a dialogue FSM transition.
+///
+/// This macro expands to a [`dptree::Handler`] that filters your dialogue
+/// state: if the state enumeration is of a certain variant, the execution
+/// continues; otherwise, `dptree` will try the next branch.
+///
+/// Variants can take the following forms:
+///
+///  - `State::MyVariant` for empty variants;
+///  - `State::MyVariant(param1, ..., paramN)` for function-like variants;
+///  - `State::MyVariant { param1, ..., paramN }` for `struct`-like variants.
+///
+/// In the first case, this macro results in a simple [`dptree::filter`]; in the
+/// second and third cases, this macro results in [`dptree::filter_map`] that
+/// passes the payload of `MyVariant` to the next handler if the match occurs.
+/// (This next handler can be an endpoint or a more complex one.) The payload
+/// format depend on the form of `MyVariant`:
+///
+///  - For `State::MyVariant(param)`, the payload is `param`.
+///  - For `State::MyVariant(param1, ..., paramN)`, the payload is `(param1,
+///    ..., paramN)` (where `N` > 1).
+///  - For `State::MyVariant { param1, ..., paramN }`, the payload is `(param1,
+///    ..., paramN)`.
+///
+/// ## Dependency requirements
+///
+///  - Your dialogue state enumeration `State`.
+#[macro_export]
+macro_rules! handler {
+    ($($variant:ident)::*) => {
+        dptree::filter(|state| matches!(state, $($variant)::*))
+    };
+    ($($variant:ident)::* ($param:ident)) => {
+        dptree::filter_map(|state| match state {
+            $($variant)::*($param) => Some($param),
+            _ => None,
+        })
+    };
+    ($($variant:ident)::* ($($param:ident),+ $(,)?)) => {
+        dptree::filter_map(|state| match state {
+            $($variant)::*($($param),+) => Some(($($param),+ ,)),
+            _ => None,
+        })
+    };
```

Comment:
Huh, this is interesting behavior. I would expect `handler![State(a)]` and `handler![State(a,)]` to be equivalent, but under this implementation they differ — the former provides `A`, while the latter provides `(A,)` (where `A` is the type of `a`). This is especially confusing given that the struct one (`handler![State { a }]`) always provides `(A,)`.

I am not sure if this is a good decision or not.

Reply:
What if we put `$(,)?` after `($param:ident)` in the first macro arm?

Reply:
Tbh, I don't know. This would make `handler![State(a)]` and `handler![State(a,)]` equivalent, which is good... This also would match the documentation:
> - For `State::MyVariant(param1, ..., paramN)`, the payload is `(param1, ..., paramN)` **(where `N` > 1)**

At the same time it is still confusing in the sense that it works differently for N = 1 and N > 1. But then if we make `handler![State(a)]` provide `(A,)` it is probably even more confusing (from my experience it can be hard to understand the difference between `(A,)` and `A`, especially for beginners).

Reply:
I guess it doesn't really matter what approach we'll take, the confusion will stay. I think we just need clear documentation for whatever choice we make.

I would suggest the following:
- `handler![State(a)]`, `handler![State(a,)]`, `handler![State { a }]` and `handler![State { a,}]` provide `A` (it's more convenient than using `(A,)`)

Reply:
Yes, actually this is the case. `MyVariant(A)` transforms into `A` for the sake of simplicity, to avoid the perplexing `(A,)` syntax. This is, in fact, backwards-compatible with the previous implementation, `#[derive(DialogueState)]`.

Reply:
Makes sense, yes. I'll fix it too.

Reply:
Or maybe we can forbid `handler![State(a,)]` and `handler![State { a, }]` at all (using `compiler_error!("Remove the trailing comma (','), it is prohibited for single-field states")`). This way every time there is a `,` it's a tuple, but also there is no confusion for the single element.

Reply:
I don't like forbidding a trailing comma because it makes macro usage inconsistent. E.g., imagine you are generating `teloxide::handler!` inside another macro; then you should concern about putting the trailing comma after enumeration variant parameters or not, which cultivates additional hassle.

Reply:
That's the point!

If you are generating `teloxide::handler!` inside another macro, you need to check for `N = 1` anyway, because it provides `A` and not `(A,)`. IMO it's better to fail early, than to compile and then fail because user macro expected `(A,)` but `A` was provided.

Reply:
Or we can make `State(A,)` and `State { A, }` generate `(A,)`, but also make `State(A)` and `State { A }` generate `A`. So if a user macro generating `teloxide::handler!` places the trailing comma after a parameter list, it can be totally sure that `teloxide::handler!` passes a tuple to its handlers. But again, this would introduce some confusion with `State(A)` and `State(A, )` and likewise with curly braces...

Reply:
On the other hand, it might be natural to conclude that `State(A)` passes a single type `A` but `State(A,)` passes `(A,)` I think?

Reply:
I don't know. `State(a)` vs `State(a,)` is indeed similar to `(a)` vs `(a,)`, but this is subtle.

Reply:
Well I think it's the most adequate solution for now. `dptree` would anyway show the list of available types on panic (during bot testing). Now I'm going to introduce the necessary changes and request your review again.